ts_algorithms/fdk.py

```diff
@@ -0,0 +1,129 @@
+import numpy as np
+import torch
+import tomosipo as ts
+from .fbp import fbp
+
+
+def fdk_weigh_projections(A, y, src_rot_center_dist):
+    # The distance of each pixel to the source is calulated in a coordinate
+    # system based on the normalized u, v and normal vectors of the detector
+    # because that way the distance along the v axis is constant along each
+    # row and the distance along the u axis is constant along each column
+    # resulting in fewer computations
+
+    src_to_det = torch.from_numpy(A.range.det_pos[0] - A.range.src_pos[0])
+    src_det_dist = torch.norm(src_to_det)
+    # Load u, v vectors:
+    det_u = torch.from_numpy(A.range.det_u[0])
+    det_v = torch.from_numpy(A.range.det_v[0])
+    # Calculate u, v vector lengths:
+    det_u_norm = torch.norm(det_u)
+    det_v_norm = torch.norm(det_v)
+    # Transform source to detector center vector to the u, v system
+    src_det_u_dist = torch.dot(src_to_det, det_u) / det_u_norm
+    src_det_v_dist = torch.dot(src_to_det, det_v) / det_v_norm
+
+    # calculate the distance of each row and column of pixels along the u or v axis
+    u_num_pixels = A.range_shape[2]
+    u_space = (torch.arange(u_num_pixels, dtype=torch.float64)+0.5-(u_num_pixels/2))
+    v_num_pixels = A.range_shape[0]
+    v_space = (torch.arange(v_num_pixels, dtype=torch.float64)+0.5-(v_num_pixels/2))
+
+    # Calculate the minimal source to detector distance divided by the
+    # pixel to detector distance for each pixel. The pixel to detector
+    # distances are calculated using the pythagorean theorem
+    u_pos_squared = (u_space * det_u_norm + src_det_u_dist)**2
+    v_pos_squared = (v_space * det_v_norm + src_det_v_dist)**2
+
+    weights_mat = src_det_dist / torch.sqrt(
+          u_pos_squared[None, :] + v_pos_squared[:, None] + src_det_dist**2
+    )
+
+    # Multiply with extra scaling factor to account for detector distance
+    weights_mat *= (src_rot_center_dist / src_det_dist)
+    weights_mat = weights_mat.float().to(y.device)
+
+    return y * weights_mat[:, None, :]
+
+
+def fit_src_rot_center_dist(A):
+    vg = A.domain
+    pg = A.range
+
+    T = ts.from_perspective(
+        pos=pg.det_pos,
+        w=pg.det_v/np.linalg.norm(pg.det_v[0]),
+        v=pg.det_normal/np.linalg.norm(pg.det_normal[0]),
+        u=pg.det_u/np.linalg.norm(pg.det_u[0])
+    )
```

Comment:
@ahendriksen This part could be made even shorter by adding a Transform.normalize_scale() function that normalizes the w, v and u vectors to 1